Improve defaults in x.py

jyn514 opened the pull request:
- Make the default stage dependent on the subcommand
- Don't build stage1 rustc artifacts with x.py build --stage 1. If this is what you want, use x.py build --stage 2 instead, which gives you a working libstd.
- Change default debuginfo when debug = true from 2 to 1

I tried to fix CI to use `--stage 2` everywhere it currently has no stage, but I might have missed a spot.
This does not update much of the documentation - most of it is in https://github.com/rust-lang/rustc-dev-guide/ or https://github.com/rust-lang/rust-forge and will need a separate PR.

See individual commits for a detailed rationale of each change.
See also the MCP: https://github.com/rust-lang/compiler-team/issues/326

r? @Mark-Simulacrum , but anyone is free to give an opinion.

jyn514 committed:
Don't duplicate builder code

- Add Builder::new_internal

jyn514 committed:
Make the default stage dependent on the subcommand

 ### x.py build/test: stage 1

I've seen very few people who actually use full stage 2 builds on purpose. These compile rustc and libstd twice and don't give you much more information than a stage 1 build (except in rare cases like https://github.com/rust-lang/rust/pull/68692#discussion_r376392145). For new contributors, this makes the build process even more daunting than it already is. As long as CI is changed to use `--stage 2` I see no downside here.

 ### x.py bench/dist/install: stage 2

These commands have to do with a finished, optimized version of rustc. It seems very rare to want to use these with a stage 1 build.

 ### x.py doc: stage 0

Normally when you document things you're just fixing a typo. In this case there is no need to build the whole rust compiler, since the documentation will usually be the same when generated with the beta compiler or with stage 1.

Note that for this release cycle only there will be a significant different between stage0 and stage1 docs: https://github.com/rust-lang/rust/pull/73101. However most of the time this will not be the case.

jyn514 committed:
Don't build rustc without std

- Set rustc to build only when explicitly asked for

This allows building the stage2 rustc artifacts, which nothing depends
on.

Previously the behavior was as follows (where stageN <-> stage(N-1) artifacts, except for stage0 libstd):

- `x.py build --stage 0`:
  - stage0 libstd
  - stage1 rustc (but without putting rustc in stage0/)

This leaves you without any rustc at all except for the beta compiler
(https://github.com/rust-lang/rust/issues/73519). This is never what you want.

- `x.py build --stage 1`:
  - stage0 libstd
  - stage1 rustc
  - stage1 libstd
  - stage1 rustdoc
  - stage2 rustc

This leaves you with a broken stage2 rustc which doesn't even have
libcore and is effectively useless. Additionally, it compiles rustc
twice, which is not normally what you want.

- `x.py build --stage 2`:
  - stage0 libstd
  - stage1 rustc
  - stage1 libstd
  - stage2 rustc
  - stage2 rustdoc and tools

This builds all tools in release mode. This is the correct usage for CI,
but takes far to long for development.

Now the behavior is as follows:

- `x.py build --stage 0`:
  - stage0 libstd

This is suitable for contributors only working on the standard library,
as it means rustc never has to be compiled.

- `x.py build --stage 1`:
  - stage0 libstd
  - stage1 rustc
  - stage1 libstd
  - stage1 rustdoc

This is suitable for contributors working on the compiler. It ensures
that you have a working rustc and libstd without having to pass
`src/libstd` in addition.

- `x.py build --stage 2`:
  - stage0 libstd
  - stage1 rustc
  - stage1 libstd
  - stage2 rustc
  - stage2 libstd
  - stage2 rustdoc

This is suitable for debugging errors which only appear with the stage2
compiler.

- `x.py build --stage 2 src/libstd src/rustc`
  - stage0 libstd
  - stage1 rustc
  - stage1 libstd
  - stage2 rustc
  - stage2 libstd
  - stage2 rustdoc, tools, etc.
  - stage2 rustc artifacts ('stage3')

This is suitable for CI, which wants all tools in release mode.
However, most of the use cases for this should use `x.py dist` instead,
which builds all the tools without each having to be named individually.

jyn514 committed:
Change debuginfo to default to 1 if `debug = true` is set

From [a conversation in discord](https://discordapp.com/channels/442252698964721669/443151243398086667/719200989269327882):

> Linking seems to consume all available RAM, leading to the OS to swap memory to disk and slowing down everything in the process
Compiling itself doesn't seem to take up as much RAM, and I'm only looking to check whether a minimal testcase can be compiled by rustc, where the runtime performance isn't much of an issue

> do you have debug = true or debuginfo-level = 2 in config.toml?
> if so I think that results in over 2GB of debuginfo nowadays and is likely the culprit
> which might mean we're giving out bad advice :(

Anecdotally, this sped up my stage 1 build from 15 to 10 minutes.

This still adds line numbers, it only removes variable and type information.

- Improve wording for debuginfo description

Co-authored-by: Teymour Aldridge <42674621+teymour-aldridge@users.noreply.github.com>

jyn514 commented:
I would also like to add an option to build rustc without libstd at some point. This is not currently possible: `build --stage 0` will compile rustc but won't put it in path (https://github.com/rust-lang/rust/issues/73519) and I think the same is true for `--stage 1`.

teymour-aldridge commented on config.toml.example:
```diff
@@ -337,7 +337,10 @@
 #
-# Defaults to 2 if debug is true
+# Note that debuginfo-level = 2 generates several gigabytes of debuginfo
+# and will slow down linking significantly.
```
Maybe better?
```suggestion
# and will slow down the linking process significantly.
```

pickfire replied:
How about this?

```
# and slows linking significantly.
```

jyn514 commented:
Looking at https://forge.rust-lang.org/compiler/mcp.html this should definitely be an MCP so marking as draft until I've gone through that process.

pickfire commented on src/bootstrap/builder.rs:
```diff
@@ -530,6 +520,18 @@ impl<'a> Builder<'a> {
         Some(help)
     }
 
+    fn get_builder(build: &Build, kind: Kind, paths: Vec<PathBuf>) -> Builder<'_> {
+        Builder {
+            build,
+            top_stage: build.config.stage.unwrap_or(1),
+            kind,
+            cache: Cache::new(),
+            stack: RefCell::new(Vec::new()),
+            time_spent_on_dependencies: Cell::new(Duration::new(0, 0)),
+            paths,
+        }
+    }
+
```
Why not implement `Default` instead of having this?

jyn514 replied:
Is there a sane default for Build, Kind, and Paths? I don't want to implement Default if the default value isn't usable on its own.

pickfire replied:
I mean one could use `{ build, kind, paths, ..Default::default() }`.

jyn514 replied:
Right, but you can't construct `Default::default()` because there is no `impl Default for Build`.

petrochenkov commented:
The current defaults (except for the default stage) minimize the edit-build-test cycle, i.e. serve the use case "build rustc as fast as possible to run tests".

The new defaults are debug-oriented - you have some issue and build the compiler with extra info, spending extra time on producing debuginfo, assertions and logging.

The first scenario is much more common from my experience.
So I use the current default settings most of the time with `debuginfo-level = 1` being the only exceptions (it's very cheap and doesn't cause any observable slowdown).

jyn514 commented:
> The current defaults (except for the default stage) minimize the edit-build-test cycle, i.e. serve the use case "build rustc as fast as possible to run tests".
> 
> The new defaults are debug-oriented - you have some issue and build the compiler with extra info, spending extra time on producing debuginfo, assertions and logging.
> 
> The first scenario is much more common from my experience.
> So I use the current default settings most of the time with `debuginfo-level = 1` being the only exceptions (it's very cheap and doesn't cause any observable slowdown).

The `debug = true` change is the one I feel the least strongly about. If you think that should be kept as `debug = false` by default I'm fine with that as long as there's a suggestion in rustc-dev-guide to switch it on. I will note though that the people with the use case to "build rustc as fast as possible to run tests" are probably experienced contributors and know to look for `debug = false` in config.toml.

What do you think about the rest of the changes?

CAD97 commented:
> I will note though that the people with the use case to "build rustc as fast as possible to run tests" are probably experienced contributors

I'd argue the exact opposite. People new to hacking on rustc are the most likely to use an unqualified `./x.py test` in order to test their changes. And they definitely just want their test results, and don't really care what's done to get to that point.

joshtriplett commented:
I would personally like to have `debug = false` by default` as well. I think it makes sense to provide better build performance out of the box; most people won't run rustc through a debugger or similar.

jyn514 commented:
To be clear, `debug = true` does not mean you can run rustc through a debugger, for that you need `debuginfo = 2`. Instead it adds line numbers to backtraces, compiles debug assertions, and emits debug logging. Without debug = true, RUST_LOG=debug has exactly the same effect as RUST_LOG=info, which can be very confusing for new contributors.

joshtriplett commented:
@jyn514 Ah, I see; that seems reasonable. Thanks for the explanation.

jyn514 commented:
I removed `debug = true` from the MCP, see https://github.com/rust-lang/compiler-team/issues/326#issuecomment-653537566 for discussion.

pickfire commented:
> To be clear, debug = true does not mean you can run rustc through a debugger, for that you need debuginfo = 2. Instead it adds line numbers to backtraces, enables debug assertions, and emits debug logging. Without debug = true, RUST_LOG=debug has exactly the same effect as RUST_LOG=info, which can be very confusing for new contributors.

The worst thing I did was changing that, I need to rebuild (which took me a day) just to flip the debug to true.

jyn514 commented:
The failure is reproducible with
```
$ ./x.py test --doc src/tools/rustdoc
   Doc-tests rustdoc

running 1 test
F
failures:

---- html/markdown.rs - html::markdown (line 5) stdout ----
Test executable failed (exit code 127).

stderr:
/tmp/rustdoctestmt2Lhu/rust_out: error while loading shared libraries: libtest-762795f83cf44985.so: cannot open shared object file: No such file or directory
```
Investigating now.

jyn514 committed:
Fix most bootstrap tests

Uses --stage 2 for all the existing tests

jyn514 committed:
Move tests into a submodule

jyn514 committed:
Add tests for the new behavior

- Only set stage 2 in dist tests
- Add test for `x.py doc` without args
- Add test for `x.py build` without args
- Add test for `x.py build --stage 0`

jyn514 commented:
Yay bootstrap tests are finally passing 🎉 next to change CI to use the explicit names for things instead of relying on the defaults to be correct.

jyn514 committed:
Use --stage 2 explicitly in CI

- expand yaml anchors
- don't use --stage 2 for dist; that's already the default

jyn514 commented:
CI is passing and using stage 2 :tada: The MCP still has 4 days until it's officially accepted, but this is ready for review.

petrochenkov commented:
@jyn514 
Could you split the changes for different defaults into separate PRs? (One for stage 2, one for debuginfo, etc.)
Right now this PR does many different things, moves code around and doesn't maintain git history.
The PR description also went out of sync with its contents.

jyn514 commented:
@petrochenkov I updated the description and split this into fewer commits, each of which does one thing. Is it easier to review now? I also dropped the `debug = true` change which snuck in by accident.

I can split the stage changes into a separate PR if you like but the rest are too small IMO to need a PR each (e.g.  https://github.com/rust-lang/rust/pull/73964/commits/34bd8740b40618fece6e69b71cfea5eb92307aaa)

jyn514 commented:
>  Right now this PR does many different things, moves code around and doesn't maintain git history.

The only code moved around are the tests in `src/bootstrap/builder/tests.rs`. If you view with whitespace changes disabled there should not be very many changes there either.

jyn514 commented:
@Mark-Simulacrum I added a test for stage0, but the behavior in the tests doesn't seem to match what I see with x.py. Here's `x.py build --stage 0`:

```
Updating only changed submodules
Submodules updated in 0.01 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.11s
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
Build completed successfully in 0:00:00
```

which as you can see doesn't touch rustdoc at all. But if I change the assertion in the tests to `assert!(builder.cache.all::<tool::Rustdoc>().is_empty());`, it gives an assertion failure. Do you know what's going on?

jyn514 commented:
Ah that's the beta rustdoc. So it doesn't need to be rebuilt. I'll add a comment to that effect.

bors commented:
:umbrella: The latest upstream changes (presumably #74461) made this pull request unmergeable. Please resolve the merge conflicts.

jyn514 commented:
@bors try

bors commented:
:hourglass: Trying commit d9b17661463fb2885d281be951f578d87eaa2b23 with merge b4557f54c17832c833fa0656e4b14d71da178ac2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d9b17661463fb2885d281be951f578d87eaa2b23","merge_sha":"b4557f54c17832c833fa0656e4b14d71da178ac2"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/884338350), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35072)
Build commit: b4557f54c17832c833fa0656e4b14d71da178ac2 (`b4557f54c17832c833fa0656e4b14d71da178ac2`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35072","checks-actions":"https://github.com/rust-lang-ci/rust/runs/884338350"},"merge_sha":"b4557f54c17832c833fa0656e4b14d71da178ac2"} -->

jyn514 commented:
Now the MCP had been accepted this is ready for review :)

Mark-Simulacrum commented on src/bootstrap/builder.rs:
```diff
@@ -534,6 +525,18 @@ impl<'a> Builder<'a> {
         Some(help)
     }
 
+    fn get_builder(build: &Build, kind: Kind, paths: Vec<PathBuf>) -> Builder<'_> {
```
```suggestion
    fn new(build: &Build, kind: Kind, paths: Vec<PathBuf>) -> Builder<'_> {
```

Will need some changes to call sites as well.

jyn514 replied:
There is already a [`Builder::new`](https://github.com/rust-lang/rust/pull/73964/files/b8f10217fe56c3b81894681fda948ccb70ce13a2#diff-4bc384d35e57dc577b9378d4d733d7c3R540). Maybe this could be `new_internal()`?

Mark-Simulacrum replied:
Yeah that seems fine. Would be okay with renaming the Builder::new as well, whatever works for you -- I don't like get_builder as it feels global-y.

Mark-Simulacrum reviewed (commented):
> See individual commits for a detailed rationale of each change.

Thank you so much for this! It'll be super useful in the future and was quite helpful when reviewing code.

I share your concern about missing some location where we're currently passing `--stage 2` implicitly -- I wonder if you'd be up for adding an assert to the flag parsing (or elsewhere, though we should be careful to not break bootstrap's own tests) that if the [in CI check](https://github.com/rust-lang/rust/blob/efc02b03d18b0cbaa55b1e421d792f70a39230b2/src/bootstrap/util.rs#L254) indicates that we're in CI, we are passing --stage 2 or that it'll be set implicitly? That way we can have some confidence that we're not missing anything.

I think this is pretty much ready to go with the check I describe added and a nit fix.

jyn514 commented:
> I share your concern about missing some location where we're currently passing --stage 2 implicitly -- I wonder if you'd be up for adding an assert to the flag parsing (or elsewhere, though we should be careful to not break bootstrap's own tests) that if the in CI check indicates that we're in CI, we are passing --stage 2 or that it'll be set implicitly?

I like this idea, but I'm not sure how to implement it. In [`sanity::check`](https://github.com/rust-lang/rust/blob/efc02b03d18b0cbaa55b1e421d792f70a39230b2/src/bootstrap/sanity.rs#L64), we have access to `Build` but not to `Builder`, so `stage` is still an `Option<u32>`. Maybe I can assert that it has to be set explicitly? `if build.subcommand == Test { assert!(build.stage == Some(2)); }`?

jyn514 committed:
Add assert that tests happen with stage 2 in CI

- Use stage 2 for makefile
- Move assert to builder
- Don't add an assert for --help
- Allow --stage 0 if passed explicitly
- Don't assert defaults during tests

Otherwise it's impossible to test the defaults!

Mark-Simulacrum commented on src/bootstrap/builder.rs:
```diff
@@ -535,6 +526,31 @@ impl<'a> Builder<'a> {
+                Kind::Check | Kind::Clippy | Kind::Fix | Kind::Run | Kind::Format => 0,
+            }
+        };
+        Builder {
```
I would expect the assert to go here I think -- does that work? Something like `if on_ci { assert_eq!(top_stage, 2); }`. Obviously we might have fallout and need to add some more "oh `x.py check` is okay" or so, I'm fine with that.

jyn514 replied:
Oh that's a much better idea 😆 I implemented that.

jyn514 commented:
Good idea with the assert - it turns out that I completely missed `src/bootstrap/mk/Makefile.in`. This should be ready for another round of review.

jyn514 committed:
Use exhaustive match for assert

Mark-Simulacrum commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 311eeedd96240d251044548b6acbc6a79f5adcd0 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 311eeedd96240d251044548b6acbc6a79f5adcd0 -->
<!-- homu: {"type":"Approved","sha":"311eeedd96240d251044548b6acbc6a79f5adcd0","approver":"Mark-Simulacrum"} -->

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout sane-defaults` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self sane-defaults --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging src/ci/docker/host-x86_64/wasm32/Dockerfile
CONFLICT (content): Merge conflict in src/ci/docker/host-x86_64/wasm32/Dockerfile
Auto-merging src/ci/docker/host-x86_64/test-various/Dockerfile
Auto-merging src/ci/docker/host-x86_64/mingw-check/Dockerfile
CONFLICT (content): Merge conflict in src/ci/docker/host-x86_64/mingw-check/Dockerfile
Auto-merging src/bootstrap/mk/Makefile.in
Auto-merging src/bootstrap/compile.rs
Auto-merging src/bootstrap/builder/tests.rs
CONFLICT (content): Merge conflict in src/bootstrap/builder/tests.rs
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #73265) made this pull request unmergeable. Please resolve the merge conflicts.

jyn514 commented:
@bors r=Mark-Simulacrum rollup=iffy

bors commented:
:pushpin: Commit a5337d668c4976fed89551f94b03b2903f8f947c has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum a5337d668c4976fed89551f94b03b2903f8f947c -->
<!-- homu: {"type":"Approved","sha":"a5337d668c4976fed89551f94b03b2903f8f947c","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit a5337d668c4976fed89551f94b03b2903f8f947c with merge 621a543f230d0562a7a6b0a2e481d466fd9738e0...
<!-- homu: {"type":"BuildStarted","head_sha":"a5337d668c4976fed89551f94b03b2903f8f947c","merge_sha":"621a543f230d0562a7a6b0a2e481d466fd9738e0"} -->

rust-log-analyzer commented:
The job `x86_64-gnu-tools` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/917342767) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/621a543f230d0562a7a6b0a2e481d466fd9738e0/checks/917342767/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/621a543f230d0562a7a6b0a2e481d466fd9738e0/checks/917342767/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
== clock drift check ==
  local time: Tue Jul 28 03:40:57 UTC 2020
  network time: Tue, 28 Jul 2020 03:40:57 GMT
== end clock drift check ==
##[error]Process completed with exit code 1.
Terminate orphan process: pid (2797) (node)
Terminate orphan process: pid (2831) (python)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/917352048)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/917352048","builder_name":"checks-actions"} -->

jyn514 committed:
Fix bad rebase

jyn514 commented:
@bors r=Mark-Simulacrum rollup=iffy

bors commented:
:pushpin: Commit c4c6453b7ba42c5114e875908eeaf510398d2f38 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum c4c6453b7ba42c5114e875908eeaf510398d2f38 -->
<!-- homu: {"type":"Approved","sha":"c4c6453b7ba42c5114e875908eeaf510398d2f38","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit c4c6453b7ba42c5114e875908eeaf510398d2f38 with merge 501728248fd9872b00c7564e859c1488c0a1d378...
<!-- homu: {"type":"BuildStarted","head_sha":"c4c6453b7ba42c5114e875908eeaf510398d2f38","merge_sha":"501728248fd9872b00c7564e859c1488c0a1d378"} -->

rust-log-analyzer commented:
The job `x86_64-gnu-tools` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/919041101) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/501728248fd9872b00c7564e859c1488c0a1d378/checks/919041101/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/501728248fd9872b00c7564e859c1488c0a1d378/checks/919041101/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
== clock drift check ==
  local time: Tue Jul 28 13:26:19 UTC 2020
  network time: Tue, 28 Jul 2020 13:26:19 GMT
== end clock drift check ==
##[error]Process completed with exit code 1.
Terminate orphan process: pid (2833) (node)
Terminate orphan process: pid (2867) (python)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/919056562)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/919056562","builder_name":"checks-actions"} -->

jyn514 committed:
Use --stage 2 in checktools

- Remove useless --stage 2 argument to checktools.sh
- Fix help text for expand-yaml-anchors (it had a typo)

jyn514 commented:
@bors r=Mark-Simulacrum rollup=iffy

bors commented:
:pushpin: Commit da40cf81e63d04dced9d943743bb2a58801916e7 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum da40cf81e63d04dced9d943743bb2a58801916e7 -->
<!-- homu: {"type":"Approved","sha":"da40cf81e63d04dced9d943743bb2a58801916e7","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit da40cf81e63d04dced9d943743bb2a58801916e7 with merge 7b3a7819371cef92a187e9bac8f7810ccde15216...
<!-- homu: {"type":"BuildStarted","head_sha":"da40cf81e63d04dced9d943743bb2a58801916e7","merge_sha":"7b3a7819371cef92a187e9bac8f7810ccde15216"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/919471605), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35654)
Approved by: Mark-Simulacrum
Pushing 7b3a7819371cef92a187e9bac8f7810ccde15216 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35654","checks-actions":"https://github.com/rust-lang-ci/rust/runs/919471605"},"merge_sha":"7b3a7819371cef92a187e9bac8f7810ccde15216"} -->

matklad commented:
I’ve just learned about this from 1.47 relnotes. This might mean that this might need some additional advertising for existing contributors to take advantage of. “How rustc stages work” might be an interesting post for inside Rust blog :)

lzutao commented:
Joshua is making a post: https://github.com/rust-lang/blog.rust-lang.org/pull/666
